File: 3rd year/PDP/Lab/Lab2/src/java/Notifier.java
package java;

import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

public class Notifier<T> {
    private T data = null;
    private final ReentrantLock lock = new ReentrantLock();
    private final Condition condition = lock.newCondition();
    private final boolean withLogs;

    public Notifier(boolean withLogs) {
        this.withLogs = withLogs;
    }

    public Notifier() {
        this.withLogs = false;
    }

    public void notify(T data) {
        lock.lock();

        try {
            while (this.data != null) {
                    condition.await();
            }

            this.data = data;
            if (withLogs) {
                System.out.println("Notifying data: " + data);
            }

            condition.signal();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        lock.unlock();
    }

    public T waitForData() {
        lock.lock();

        try {
            while (this.data == null) {
                condition.await();
            }

            var data = readData();
            if (withLogs) {
                System.out.println("Waited for data: " + data);
            }

            condition.signal();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        lock.unlock();

        return data;
    }

    private T readData() {
        var data = this.data;

        this.data = null;

        return data;
    }
}

File: 3rd year/PDP/Lab/Lab2/src/Notifier.kt
import java.util.concurrent.locks.ReentrantLock

class Notifier<T>(private val withLogs: Boolean = false) {
    private var data: T? = null
    private val lock = ReentrantLock()
    private val condition = lock.newCondition()

    fun notify(data: T) {
        lock.lock()

        while (this.data != null) {
            condition.await()
        }

        this.data = data
        if (withLogs) {
            println("Notifying data: $data")
        }
        condition.signal()

        lock.unlock()
    }

    fun waitForData(): T {
        lock.lock()

        while (data == null) {
            condition.await()
        }

        val data = readData()
        if (withLogs) {
            println("Waited for data: $data")
        }
        condition.signal()

        lock.unlock()

        return data!!
    }

    private fun readData(): T? {
        val data = this.data

        this.data = null

        return data
    }
}